src/mbgl/renderer/layers/render_location_indicator_layer.cpp

```diff
@@ -0,0 +1,800 @@
+#include <array>
+#include <mapbox/cheap_ruler.hpp>
+#include <mbgl/gfx/backend_scope.hpp>
+#include <mbgl/gfx/renderer_backend.hpp>
+#include <mbgl/gl/context.hpp>
+#include <mbgl/gl/renderable_resource.hpp>
+#include <mbgl/map/transform_state.hpp>
+#include <mbgl/platform/gl_functions.hpp>
+#include <mbgl/renderer/bucket.hpp>
+#include <mbgl/renderer/layers/render_location_indicator_layer.hpp>
+#include <mbgl/renderer/paint_parameters.hpp>
+#include <mbgl/style/layers/location_indicator_layer.hpp>
+#include <mbgl/style/layers/location_indicator_layer_impl.hpp>
+#include <mbgl/style/layers/location_indicator_layer_properties.hpp>
+#include <mbgl/util/mat4.hpp>
+
+#include <mapbox/eternal.hpp>
+#include <mbgl/gl/context.hpp>
+#include <mbgl/gl/defines.hpp>
+#include <mbgl/gl/texture.hpp>
+#include <mbgl/gl/texture_resource.hpp>
+#include <mbgl/gl/types.hpp>
+#include <mbgl/platform/gl_functions.hpp>
+#include <mbgl/renderer/image_manager.hpp>
+
+using namespace mbgl::platform;
+namespace mbgl {
+
+struct LocationIndicatorRenderParameters {
+    LocationIndicatorRenderParameters() = default;
+    explicit LocationIndicatorRenderParameters(const TransformParameters& tp) : state(&tp.state) {}
+    LocationIndicatorRenderParameters(const LocationIndicatorRenderParameters& o) = default;
+    LocationIndicatorRenderParameters& operator=(const LocationIndicatorRenderParameters& o) = default;
+
+    double width;
+    double height;
+    double latitude;
+    double longitude;
+    double zoom;
+    double bearing;
+    double pitch;
+    double fieldOfView;
+    std::array<double, 16> projectionMatrix;
+    const TransformState* state = nullptr;
+    ImageManager* imageManager = nullptr;
+    // some testing defaults, for before it gets updated via props
+    double puckBearing = 0.0;
+    LatLng puckPosition = {0, 0};
+    double errorRadiusMeters;
+    mbgl::Color errorRadiusColor{0, 0, 0, 0};
+    mbgl::Color errorRadiusBorderColor{0, 0, 0, 0};
+    int puckSizePx = 0;
+    int puckHatSizePx = 0;
+    int puckShadowSizePx = 0;
+    float puckLayersDisplacement = 0;
+    float perspectiveCompensation = 0;
+    std::string puckImagePath;
+    std::string puckShadowImagePath;
+    std::string puckHatImagePath;
+};
+
+class RenderLocationIndicatorImpl {
+protected:
+    struct vec2 {
```

Comment:
looks like a generic class, don't we already have it (or shall be put to a generic place)?

Reply:
the closest seems to be Point<T>, missing methods such as normalize(d) and length